feat: add new CronTrigger task type

zevaryx opened the pull request:
## Pull Request Type
<!-- Check the appropriate option -->

- [x] Feature addition
- [ ] Bugfix
- [ ] Documentation update
- [ ] Code refactor
- [ ] Tests improvement
- [ ] CI/CD pipeline enhancement
- [ ] Other: [Replace with a description]

## Description
<!-- Provide a clear and concise description of the purpose of this PR and why it should be merged -->
<!-- If your code adds or changes features, a usage example would be helpful -->
Adds a new Task Trigger, CronTrigger. Pass a cron string to it (i.e. `0 0 * * *`), and it'll run tasks based on that cron string

## Changes
<!-- List the changes you have made in a bullet-point format -->
- Add `croniter` requirement
- Add `CronTrigger` task trigger

## Related Issues
<!-- If this PR is related to any open issues, please mention them using "#ISSUENUMBER" -->


## Test Scenarios
<!-- Provide clear instructions on how to test your changes, where applicable - if no tests are required, explain why -->
```py
@Task.create(CronTrigger("* * * * *"))
async def test_cron(self) -> None:
    print("I run every minute!")
```

## Python Compatibility
<!-- Testing 3.11 is not strictly required, but it would be nice if you could confirm that it works. -->
- [x] I've ensured my code works on Python `3.10.x`
- [x] I've ensured my code works on Python `3.11.x`


## Checklist
<!-- If you have not completed all of the following, there is a good chance your PR will not be merged. -->
- [ ] I've run the `pre-commit` code linter over all edited files
- [x] I've tested my changes on supported Python versions
- [x] I've added tests for my code, if applicable
- [x] I've updated / added  documentation, where applicable

zevaryx committed:
[feat] Add CronTrigger

AstreaTSS commented:
This debatably solves #1560.

zevaryx committed:
[feat] Add timezone support to CronTrigger

silasary commented:
> This debatably solves https://github.com/interactions-py/interactions.py/issues/1560.

I think it does, yeah

AstreaTSS commented on interactions/models/internal/tasks/triggers.py:
```diff
@@ -140,3 +142,20 @@ def _set_current_trigger(self) -> BaseTrigger | None:
+
+    Attributes:
+        cron str: The cron schedule, use https://crontab.guru for help
+        utc bool: Whether or not to use UTC for the time
```
Seems like you forgot to change this part 😅

eightween replied:
What about this needs to be changed?

AstreaTSS replied:
The `utc` docstring entry needs to be replaced with a `tz` entry.

eightween committed:
docs: utc -> tz rename

Signed-off-by: 18 <111544899+xvi-ii@users.noreply.github.com>

pre-commit-ci[bot] committed:
ci: correct from checks.

eightween committed:
docs: specify tz as datetime attribute

I'm aware that datetime is not what it actually is, but we shouldn't be using private data types in our docstrings. Developers will additionally be using it as a reference, not the literal representation when using

Signed-off-by: 18 <111544899+xvi-ii@users.noreply.github.com>

AstreaTSS reviewed (approved):
Don't know what's going on with the PR's diff, but looks fine otherwise.
